go/vt/vtgate/evalengine/comparisons.go

```diff
@@ -126,63 +150,104 @@ func evalResultsAreDates(l, r EvalResult) bool {
-	return sqltypes.IsDate(l.typ) && (sqltypes.IsText(r.typ) || sqltypes.IsBinary(r.typ)) ||
-		(sqltypes.IsText(l.typ) || sqltypes.IsBinary(l.typ)) && sqltypes.IsDate(r.typ)
+	return (sqltypes.IsDate(l.typ) && r.textual()) || (l.textual() && sqltypes.IsDate(r.typ))
 }
 
 func evalResultsAreDateAndNumeric(l, r EvalResult) bool {
 	return sqltypes.IsDate(l.typ) && sqltypes.IsNumber(r.typ) || sqltypes.IsNumber(l.typ) && sqltypes.IsDate(r.typ)
 }
 
-func nullSafeCoerceAndCompare(lVal, rVal EvalResult) (comp int, isNull bool, err error) {
+func evalCoerceAndCompare(lVal, rVal EvalResult) (comp int, isNull bool, err error) {
 	if lVal.collation.Collation != rVal.collation.Collation {
 		lVal, rVal, err = mergeCollations(lVal, rVal)
 		if err != nil {
 			return 0, false, err
 		}
 	}
-	return nullSafeCompare(lVal, rVal)
+	return evalCompare(lVal, rVal)
 }
 
-// For more details on comparison expression evaluation and type conversion:
-// 		- https://dev.mysql.com/doc/refman/8.0/en/type-conversion.html
-func nullSafeCompare(lVal, rVal EvalResult) (comp int, isNull bool, err error) {
-	lVal = foldSingleLenTuples(lVal)
-	rVal = foldSingleLenTuples(rVal)
+func evalCoerceAndCompareNullSafe(lVal, rVal EvalResult) (comp int, err error) {
+	if lVal.collation.Collation != rVal.collation.Collation {
+		lVal, rVal, err = mergeCollations(lVal, rVal)
+		if err != nil {
+			return 0, err
+		}
+	}
+	return evalCompareNullSafe(lVal, rVal)
+}
+
+func evalTypecheckTuples(lVal, rVal EvalResult) (bool, error) {
+	switch {
+	case lVal.typ == querypb.Type_TUPLE && rVal.typ == querypb.Type_TUPLE:
+		return true, nil
+	case lVal.typ == querypb.Type_TUPLE:
+		return false, vterrors.NewErrorf(vtrpcpb.Code_INVALID_ARGUMENT, vterrors.OperandColumns, "Operand should contain %d column(s)", len(*lVal.tuple))
+	case rVal.typ == querypb.Type_TUPLE:
+		return false, vterrors.NewErrorf(vtrpcpb.Code_INVALID_ARGUMENT, vterrors.OperandColumns, "Operand should contain %d column(s)", 1)
+	default:
+		return false, nil
+	}
+}
+
+func evalCompareNullSafe(lVal, rVal EvalResult) (cmp int, err error) {
+	tuple, err := evalTypecheckTuples(lVal, rVal)
+	if err != nil {
+		return 0, err
+	}
+	if tuple {
+		return evalCompareTuplesNullSafe(lVal, rVal)
+	}
+	if hasNullEvalResult(lVal, rVal) {
+		if lVal.typ == rVal.typ {
+			return 0, nil
+		}
+		return 1, nil
+	}
+	return evalCompare1(lVal, rVal)
+}
+
+func evalCompare(lVal, rVal EvalResult) (comp int, isNull bool, err error) {
+	tuple, err := evalTypecheckTuples(lVal, rVal)
+	if err != nil {
+		return 0, false, err
+	}
+	if tuple {
+		return evalCompareTuples(lVal, rVal)
+	}
 	if hasNullEvalResult(lVal, rVal) {
 		return 0, true, nil
 	}
+	comp, err = evalCompare1(lVal, rVal)
+	return comp, false, err
+}
+
+// For more details on comparison expression evaluation and type conversion:
+// 		- https://dev.mysql.com/doc/refman/8.0/en/type-conversion.html
+func evalCompare1(lVal, rVal EvalResult) (comp int, err error) {
```

Comment:
once the above method `evalCompare` is renamed, we can rename this to `evalCompare`